lib/polygeist/Passes/RaiseToLinalg.cpp

```diff
@@ -0,0 +1,184 @@
+#include "PassDetails.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Passes.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "polygeist/Passes/Passes.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "raise-to-affine"
+
+using namespace mlir;
+using namespace mlir::arith;
+using namespace polygeist;
+using namespace affine;
+
+namespace {
+struct RaiseAffineToLinalg : public SCFRaiseToAffineBase<RaiseAffineToLinalg> {
+  void runOnOperation() override;
+};
+} // namespace
+
+// Also want to add support for affine.for ( ) { linalg.generic } -> bigger linalg.generic
+// Also probably want to try to do { linalg.generc1(); linalg.generic2(); } -> bigger linalg.generic()
+
+/*
+
+affine.for() {
+    affine.for() {
+    } 
+    affine.for() {
+    }
+}
+
+*/
+struct Condition {
+    bool ifTrue;
+    AffineIfOp op;
+};
+struct ForOpRaising : public OpRewritePattern<affine::ForOp> {
```

Comment:
Interesting!! What would be the raising strategy here? I did some access/structural matchers a long time ago to raise affine/scf to Linalg.

Reply:
@ftynse

Reply:
https://grosser.science/static/7d02fb58ecc49e4d2097d11bc9e8152a/chelini-2021-abstraction-raising.pdf